### ISSUE TITLE
better instance Hashable IntSet?

### ISSUE THREAD

**Issue by jwaldmann** [haskell/containers#964]
cross reference to https://github.com/haskell-unordered-containers/hashable/issues/269

**Comment by treeowl** [comment]
@jwaldmann Correct.

**Comment by meooow25** [comment]
Correct, but it would be wasteful. The `Tip`s contain all the elements, it would be sufficient to just hash the contents of those (ignoring `Bin`s).

**Comment by treeowl** [comment]
> Correct, but it would be wasteful. The `Tip`s contain all the elements, it would be sufficient to just hash the contents of those (ignoring `Bin`s).

Oh yeah, I think there is redundant info there. I'd have to review the representation; it's been a little while.

**Comment by treeowl** [comment]
We could offer

```haskell
serialize :: (Word -> r -> r) -> r -> IntSet -> r
```

This could be used to hash efficiently, with `serialize` written so it can inline. What would be a good matching `deserialize` type though?

**Comment by treeowl** [comment]
I doubt that's quite what we'd want for `Ord`, but maybe close. The use case is that I'd be more comfortable with "publicly" exporting serialize/deserialize than a function that's only good for writing the `Hashable` instance.

**Comment by meooow25** [comment]
I think we should not mix up serialization and hashing.  
Serialization typically requires the size before the elements, which means we must make two traversals for IntMap/IntSet.  
Hashing also wants the size for a good hash, but it is fine to add it at the end, which means we can be more efficient and calculate the hash and size in one traversal.
